component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/handler/DPoPAuthenticationHandler.java

```diff
@@ -0,0 +1,390 @@
+                authenticationContext.addParameter(OAUTH2_ALLOWED_SCOPES, responseDTO.getScope());
+                authenticationContext.addParameter(OAUTH2_VALIDATE_SCOPE,
+                        AuthConfigurationUtil.getInstance().isScopeValidationEnabled());
+                String serviceProvider = null;
+                try {
+                    serviceProvider =
+                            OAuth2Util.getServiceProvider(clientApplicationDTO.getConsumerKey()).getApplicationName();
+                } catch (IdentityOAuth2Exception e) {
+                    log.error("Error occurred while getting the Service Provider by Consumer key: "
+                            + clientApplicationDTO.getConsumerKey());
+                }
+
+                String serviceProviderTenantDomain = null;
+                try {
+                    serviceProviderTenantDomain =
+                            OAuth2Util.getTenantDomainOfOauthApp(clientApplicationDTO.getConsumerKey());
+                } catch (InvalidOAuthClientException | IdentityOAuth2Exception e) {
+                    log.error("Error occurred while getting the OAuth App tenantDomain by Consumer key: "
+                            + clientApplicationDTO.getConsumerKey());
+                }
+
+                if (serviceProvider != null) {
+                    authenticationContext.addParameter(SERVICE_PROVIDER, serviceProvider);
+                    if (serviceProviderTenantDomain != null) {
+                        authenticationContext.addParameter(SERVICE_PROVIDER_TENANT_DOMAIN, serviceProviderTenantDomain);
+                    }
+
+                    MDC.put(SERVICE_PROVIDER, serviceProvider);
+                    // Set OAuth service provider details to be consumed by the provisioning framework.
+                    setProvisioningServiceProviderThreadLocal(clientApplicationDTO.getConsumerKey(),
+                            serviceProviderTenantDomain);
+                }
+            }
+        }
+        return authenticationResult;
+    }
+
+    @Override
+    public int getPriority(MessageContext messageContext) {
+
+        return getPriority(messageContext, 24);
+    }
+
+    @Override
+    public boolean canHandle(MessageContext messageContext) {
+
+        return  isAuthHeaderMatch(messageContext, OAUTH_DPOP_HEADER);
+    }
+
+    /**
+     * Validate access token binding value.
+     *
+     * @param messageContext message context.
+     * @param tokenBinding   token binding.
+     * @param clientId       OAuth2 client id.
+     * @param accessToken    Bearer token from request.
+     * @return true if token binding is valid.
+     */
+    private boolean isTokenBindingValid(MessageContext messageContext, TokenBinding tokenBinding, String clientId,
+                                        String accessToken) {
+
+        if (tokenBinding == null || StringUtils.isBlank(tokenBinding.getBindingReference())) {
```

Comment:
Is it correct to return true if the binding value is empty?

Reply:
removed the method